hal/sai_api.h

```diff
@@ -0,0 +1,153 @@
+                                     one's clocks can be tied to the other one.
+                                     true if this endpoint it following clock from its counterpart. */
+  bool          is_receiver;    /**< true if this is a receiver. */
+  bool          is_slave;       /**< true if clocks are inputs. */
+
+  PinName       mclk;           /**< master clock pin (opional, can be NC). */
+  PinName       sd;             /**< Data pin. */
+  PinName       bclk;           /**< Bit clock pin. */
+  PinName       wclk;           /**< Word clock pin. */
+
+  bool          mclk_internal_src; /**< Set true to use internal master clock source. */
+  uint32_t      mclk_freq;      /**< Set the master clock frequency. */
+
+  sai_format_t  format;         /**< Describes the frame format. */
+} sai_init_t;
+
+/** Initialization return status.*/
+typedef enum sai_result_e {
+    SAI_RESULT_OK,                  /**< Everything went well. */
+    SAI_RESULT_DEVICE_BUSY,         /**< All endpoint for the requested type are busy. */
+    SAI_RESULT_CONFIG_UNSUPPORTED,  /**< Some requested features are not supported by this platform. */
+    SAI_RESULT_CONFIG_MISMATCH,     /**< Some requested features mismatch with the required config
+                                         (depends on underlying device state) */
+} sai_result_t;
+
+/** Delegated typedef @see target/ ** /object.h */
+typedef struct sai_s sai_t;
+
+/** SAI configuration for I2S Philips 16 bit data & word size */
+extern const sai_format_t sai_mode_i2s16;
+/** SAI configuration for I2S Philips 32 bit data & word size */
+extern const sai_format_t sai_mode_i2s32;
+/** SAI configuration for PCM 16 bit data & word size with long sync */
+extern const sai_format_t sai_mode_pcm16l;
+/** SAI configuration for PCM 16 bit data & word size with short sync */
+extern const sai_format_t sai_mode_pcm16s;
+
+/** Initialize `obj` based on `init` values.
+ * @return initialization result.
+ */
+sai_result_t sai_init(sai_t *obj, sai_init_t *init);
+
+/**
+ * Attempt to transmit a sample.
+ * @param   obj     This SAI instance.
+ * @param   sample  Sample to send.
+ * @return true if the sample was succesfully pushed to the bus (or the internal fifo).
+ */
+bool sai_write(sai_t *obj, uint32_t sample);
+
+/**
+ * Receive a sample.
+ * @param obj       This SAI instance.
+ * @param sample    Pointer to store the received sample.
+ * @return false if no sample was received or if "sample" is NULL.
+ */
+bool sai_read(sai_t *obj, uint32_t *sample);
+
+/**
+ * Attempt to transmit or receive a sample depending of the mode of this instance.
+ * @param obj       This SAI instance.
+ * @param sample    Pointer to store or fetch a sample.
+ * @return true if the/a sample was transmitted/received.
+ */
+bool sai_xfer(sai_t *obj, uint32_t *sample);
+
+/** Releases & de-initialize the referenced peripherals. */
+void sai_free(sai_t *obj);
+
+/**@}*/
+
+#ifdef __cplusplus
+}
+#endif
```

Comment:
Having multiple #endif statements, I would suggest havign a comment indicationg what they're closing.

Something line `#endif // __cplusplus`

Though off of the top of my head, I'm not sure if this is part of our style guidelines.